### PR TITLE
Add Filename and Line number to ansible_failed_task

### LINKED ISSUE #64625: Add Filename and Line number to ansible_failed_task
##### SUMMARY

I'd like to see variables added to ansible_failed_task which represent the filename and line of the task that failed.  This information is available inside the JUnit callback module already, but would be helpful to expose for better information in a rescue block notice.

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME

ansible_failed_task

##### ADDITIONAL INFORMATION

The name of the task is nice, but in large projects may not be specific enough to hunt though and find the file/line to troubleshoot.  This is even worse if the names of tasks aren't unique.  This provide something like a slack task in a rescue block the ability to inform where the fault is possible to have come from.

```yaml

```

### PR DESCRIPTION
**Pull request by Tazminia**
##### SUMMARY
Fixes #64625 

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
core

##### ADDITIONAL INFORMATION
**Default** behavior
```
(venv) tjegham@TSJ-Laptop:~/ansible$ ansible-playbook /tmp/fail.yaml
[WARNING]: No inventory was parsed, only implicit localhost is available

[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'


PLAY [localhost] **************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ********************************************************************************************************************************************************************************************
ok: [localhost]

TASK [fail] *******************************************************************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {"changed": false, "msg": "This has failed"}

PLAY RECAP ********************************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0
```
When **ANSIBLE_ALWAYS_DISPLAY_FAILED_TASK_PATH** is set to **True**
```
(venv) tjegham@TSJ-Laptop:~/ansible$ env ANSIBLE_ALWAYS_DISPLAY_FAILED_TASK_PATH=True ansible-playbook /tmp/fail.yaml
[WARNING]: No inventory was parsed, only implicit localhost is available

[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'


PLAY [localhost] **************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ********************************************************************************************************************************************************************************************
ok: [localhost]

TASK [fail] *******************************************************************************************************************************************************************************************************
task path: /tmp/fail.yaml:3
fatal: [localhost]: FAILED! => {"changed": false, "msg": "This has failed"}

PLAY RECAP ********************************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0
```
When **verbose**
```
(venv) tjegham@TSJ-Laptop:~/ansible$ ansible-playbook /tmp/fail.yaml -vv
ansible-playbook 2.10.0.dev0
  config file = None
  configured module search path = ['/home/tjegham/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/tjegham/ansible/lib/ansible
  executable location = /home/tjegham/ansible/bin/ansible-playbook
  python version = 3.6.8 (default, Oct  7 2019, 12:59:55) [GCC 8.3.0]
No config file found; using defaults
[WARNING]: No inventory was parsed, only implicit localhost is available

[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'


PLAYBOOK: fail.yaml ***********************************************************************************************************************************************************************************************
1 plays in /tmp/fail.yaml

PLAY [localhost] **************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ********************************************************************************************************************************************************************************************
task path: /tmp/fail.yaml:1
ok: [localhost]
META: ran handlers

TASK [fail] *******************************************************************************************************************************************************************************************************
task path: /tmp/fail.yaml:3
fatal: [localhost]: FAILED! => {"changed": false, "msg": "This has failed"}

PLAY RECAP ********************************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0
```

### CODE REVIEW

File: lib/ansible/plugins/doc_fragments/default_callback.py [hunk 1]
```diff
@@ -66,4 +66,15 @@ class ModuleDocFragment(object):
...
+        name: Always display failed tasks paths
+        description: 'Display failed tasks path even in non verbose mode'
+        type: bool
+        default: no
+        env:
+          - name: ANSIBLE_ALWAYS_DISPLAY_FAILED_TASK_PATH
+        ini:
+          - key: always_display_failed_task_path
+            section: defaults
+        version_added: 'x.y'
```
Comment (jborean93) [inline lib/ansible/plugins/doc_fragments/default_callback.py]:
```suggestion
        version_added: '2.10'
```

**Review by jborean93 — CHANGES_REQUESTED** [review]

File: lib/ansible/plugins/callback/default.py [hunk 2]
```diff
@@ -240,12 +243,13 @@ def _print_task_banner(self, task):
...
         self._display.banner(u"%s [%s%s]%s" % (prefix, task_name, args, checkmsg))
         if self._display.verbosity >= 2:
-            path = task.get_path()
-            if path:
-                self._display.display(u"task path: %s" % path, color=C.COLOR_DEBUG)
-
+            self._print_task_path(task.get_path())
         self._last_task_banner = task._uuid
 
+    def _print_task_path(self, path):
```
Comment (samdoran) [inline lib/ansible/plugins/callback/default.py]: I would put this method above `set_options()`. I prefer called code before the caller, which generally results in all private methods being listed before the public ones. I know there are private methods in the middle of the public ones in this class, but this new private method should go in the "correct" place.

File: lib/ansible/plugins/callback/default.py [hunk 3]
```diff
@@ -57,7 +57,8 @@
                   ('display_ok_hosts', True),
                   ('show_custom_stats', C.SHOW_CUSTOM_STATS),
                   ('display_failed_stderr', False),
-                  ('check_mode_markers', False),)
+                  ('check_mode_markers', False),
+                  ('always_display_failed_task_path', False),)
```
Comment (samdoran) [inline lib/ansible/plugins/callback/default.py]:
```suggestion
```
These options are only for backwards compatibility with older plugins. New options should not be added here.

File: lib/ansible/plugins/callback/default.py [hunk 4]
```diff
@@ -106,6 +107,8 @@ def v2_runner_on_failed(self, result, ignore_errors=False):
             self._process_items(result)
 
         else:
+            if self._display.verbosity < 2 and self.always_display_failed_task_path:
```
Comment (samdoran) [inline lib/ansible/plugins/callback/default.py]:
```suggestion
            if self._display.verbosity < 2 and self.get_option('always_display_failed_task_path'):
```

File: lib/ansible/plugins/doc_fragments/default_callback.py [hunk 5]
```diff
@@ -66,4 +66,15 @@ class ModuleDocFragment(object):
           - key: show_per_host_start
             section: defaults
         version_added: '2.9'
+      always_display_failed_task_path:
+        name: Always display failed tasks paths
+        description: 'Display failed tasks path even in non verbose mode'
```
Comment (samdoran) [inline lib/ansible/plugins/doc_fragments/default_callback.py]:
```suggestion
        description: 'When a task fails, display the path to the task file and the line number of the task.'
```

File: lib/ansible/plugins/doc_fragments/default_callback.py [hunk 6]
```diff
@@ -66,4 +66,15 @@ class ModuleDocFragment(object):
           - key: show_per_host_start
             section: defaults
         version_added: '2.9'
+      always_display_failed_task_path:
```
Comment (samdoran) [inline lib/ansible/plugins/doc_fragments/default_callback.py]:
```suggestion
      display_failed_path:
```

File: lib/ansible/plugins/doc_fragments/default_callback.py [hunk 7]
```diff
@@ -66,4 +66,15 @@ class ModuleDocFragment(object):
...
+        name: Always display failed tasks paths
```
Comment (samdoran) [inline lib/ansible/plugins/doc_fragments/default_callback.py]:
```suggestion
        name: Display path to failed tasks
```

File: lib/ansible/plugins/doc_fragments/default_callback.py [hunk 8]
```diff
@@ -66,4 +66,15 @@ class ModuleDocFragment(object):
...
+        description: 'Display failed tasks path even in non verbose mode'
+        type: bool
+        default: no
+        env:
+          - name: ANSIBLE_ALWAYS_DISPLAY_FAILED_TASK_PATH
```
Comment (samdoran) [inline lib/ansible/plugins/doc_fragments/default_callback.py]:
```suggestion
          - name: ANSIBLE_DISPLAY_FAILED_PATH
```

File: lib/ansible/plugins/doc_fragments/default_callback.py [hunk 9]
```diff
@@ -66,4 +66,15 @@ class ModuleDocFragment(object):
...
+        ini:
+          - key: always_display_failed_task_path
```
Comment (samdoran) [inline lib/ansible/plugins/doc_fragments/default_callback.py]:
```suggestion
          - key: display_failed_path
```

**Review by samdoran — CHANGES_REQUESTED** [review]
The option name is a bit too wordy. How about `display_failed_path`?

This also needs a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment) and integration tests. See [this fragment](https://github.com/ansible/ansible/blob/c2f528b7686c9573aa8493357219e0bbe4416dd3/changelogs/fragments/reboot-change-default-boot-command.yaml) as an example. See [test/integration/targets/callback_default/](https://github.com/ansible/ansible/blob/devel/test/integration/targets/callback_default/runme.sh) for existing integration tests.

**Comment by Tazminia** [comment]
@samdoran, I actually included all the changes you suggested. One odd thing however, the integration test cleanup section was not working, the created files were left undeleted until i removed the quotes around the command.

File: changelogs/fragments/64625-show-failed-task-path.yml [hunk 10]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - callback - When a task fails, display the path to the task file and the line number of the task.
```
Comment (samdoran) [inline changelogs/fragments/64625-show-failed-task-path.yml]:
```suggestion
  - default callback - add ``dispaly_failed_path`` option to display the path to the task file and the line number of the task when a task fails
```

File: test/integration/targets/callback_default/runme.sh [hunk 11]
```diff
@@ -15,6 +15,8 @@ set -eux
 
 run_test() {
 	local testname=$1
+	local TEST_FOLDER_PATH
```
Comment (samdoran) [inline test/integration/targets/callback_default/runme.sh:18]: There's no need to do this. You can simply use `$ANSIBLE_PLAYBOOK_DIR`.
```suggestion

```

File: test/integration/targets/callback_default/runme.sh [hunk 12]
```diff
@@ -15,6 +15,8 @@ set -eux
...
+	TEST_FOLDER_PATH="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
```
Comment (samdoran) [inline test/integration/targets/callback_default/runme.sh:19]:
```suggestion
	TEST_FOLDER_PATH="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
```

File: test/integration/targets/callback_default/runme.sh [hunk 13]
```diff
@@ -24,6 +26,11 @@ run_test() {
 	# Scrub deprication warning that shows up in Python 2.6 on CentOS 6
 	sed -i -e '/RandomPool_DeprecationWarning/d' "${OUTFILE}.${testname}.stderr"
 
+	# Printing task path includes absolute paths specific to the machine
+	# Replace any occurences of test folder path with TEST_FOLDER_PATH palceholder
+	sed -i -e "s#${TEST_FOLDER_PATH}#TEST_FOLDER_PATH#g" "${OUTFILE}.${testname}.stdout"
```
Comment (samdoran) [inline test/integration/targets/callback_default/runme.sh]: Only do this when needed rather than for every single test.
```suggestion
	if [[ $testname == "display_failed_path" ]]; then
		# Printing task path includes absolute paths specific to the machine
		# Replace any occurences of $ANSIBLE_PLAYBOOK_DIR with TEST_FOLDER_PATH palceholder
		sed -i -e "s#${ANSIBLE_PLAYBOOK_DIR}#TEST_FOLDER_PATH#g" "${OUTFILE}.${testname}.stdout"
	fi
```

It may also be better to split this out into a separate function and call it directly rather than putting it inside this general purpose function.

**Review by samdoran — CHANGES_REQUESTED** [review]
A few changes to the integration tests, but looks good overall.

File: test/integration/targets/callback_default/runme.sh (hunk 13 above)
Comment (Tazminia) [reply test/integration/targets/callback_default/runme.sh]: Done

File: test/integration/targets/callback_default/runme.sh (hunk 11 above)
Comment (Tazminia) [reply test/integration/targets/callback_default/runme.sh:18]: ANSIBLE_PLAYBOOK_DIR does not seem to resolve correctly in the test script. It looks like it is not set

Comment (samdoran) [reply test/integration/targets/callback_default/runme.sh:18]: `ANSIBLE_PLAYBOOK_DIR` is set by `ansible-test`, so it will be defined. Here's some of the test output showing when it is set:

```
# ansible-test integration --docker centos7 callback_default -vvv
...
Running callback_default integration test script
Preparing temporary directory: /root/ansible/test/results/.tmp/integration/callback_default-HaGZIh-ÅÑŚÌβŁÈ
Copying test/integration/targets/callback_default/ to /root/ansible/test/results/.tmp/integration/callback_default-HaGZIh-ÅÑŚÌβŁÈ/test/integration/targets/callback_default/
Copying /root/ansible/test/lib/ansible_test/_data/ansible.cfg to /root/ansible/test/results/.tmp/integration/callback_default-HaGZIh-ÅÑŚÌβŁÈ/test/integration/integration.cfg
Copying /root/ansible/test/lib/ansible_test/_data/inventory to /root/ansible/test/results/.tmp/integration/callback_default-HaGZIh-ÅÑŚÌβŁÈ/test/integration/inventory
Injecting "/tmp/python-XzS8nf-ansible/python" as a symlink to the "/usr/bin/python2.7" interpreter.
Run command: ./runme.sh -vvv
Working directory: /root/ansible/test/results/.tmp/integration/callback_default-HaGZIh-ÅÑŚÌβŁÈ/test/integration/targets/callback_default
Program found: ./runme.sh
ANSIBLE_CALLBACK_WHITELIST=junit
ANSIBLE_CONFIG=/root/ansible/test/results/.tmp/integration/callback_default-HaGZIh-ÅÑŚÌβŁÈ/test/integration/integration.cfg
ANSIBLE_DEPRECATION_WARNINGS=false
ANSIBLE_FORCE_COLOR=true
ANSIBLE_FORCE_HANDLERS=true
ANSIBLE_HOST_KEY_CHECKING=false
ANSIBLE_HOST_PATTERN_MISMATCH=error
ANSIBLE_INVENTORY=/dev/null
ANSIBLE_LIBRARY=/dev/null
ANSIBLE_PLAYBOOK_DIR=/root/ansible/test/results/.tmp/integration/callback_default-HaGZIh-ÅÑŚÌβŁÈ/test/integration/targets/callback_default
...

File: changelogs/fragments/64625-show-failed-task-path.yml [hunk 14]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - default callback - add ``dispaly_failed_path`` option to display the path to the task file and the line number of the task when a task fails
```
Comment (odyssey4me) [inline changelogs/fragments/64625-show-failed-task-path.yml:2]: s/dispaly/display/

**Comment by samdoran** [comment]
Superseded by #73260
